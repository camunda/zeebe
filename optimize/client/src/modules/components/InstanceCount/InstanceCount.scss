@use '@carbon/styles/scss/spacing';

.InstanceCount {
  display: flex;
  align-items: center;
  gap: spacing.$spacing-02;

  .countString,
  .defaultTrigger {
    font-size: 12px;
    color: var(--cds-text-secondary);
  }

  .instanceCountPopover {
    display: inline-flex;

    .defaultTrigger {
      padding: 0;
      min-height: auto;
      display: inline-block;
      text-decoration: underline;
      font-weight: bold;
    }

    .popoverContent {
      min-width: 397px;
      padding-top: 20px;

      .countString {
        font-size: 12px;
        margin-bottom: 14px;
        color: var(--grey-darken-33);
        font-weight: bold;
      }

      .filterListHeading {
        font-size: 12px;
        color: var(--grey-lighten-56);
        margin-bottom: 8px;
      }

      .FilterList {
        margin-top: 0;
        display: block;
        margin-bottom: 20px;

        &:last-child {
          margin-bottom: 0;
        }

        .listItem {
          width: 100%;
          margin: 0;
          margin-bottom: 5px;
          cursor: unset;
        }

        .ActionItem {
          min-height: 41px;

          button {
            display: none;
          }

          .parameterName {
            max-width: unset;
            overflow: unset;
          }

          .filterValues {
            width: max-content;
          }
        }

        .listItemConnector {
          display: none;
        }
      }
    }
  }
}
