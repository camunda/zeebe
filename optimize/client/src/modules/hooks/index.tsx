/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {default as useErrorHandling} from './useErrorHandling';
export {default as useUser} from './useUser';
export {default as useDocs} from './useDocs';
export {default as useUiConfig} from './useUiConfig';
