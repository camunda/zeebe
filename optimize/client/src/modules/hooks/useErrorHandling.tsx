/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useCallback, useEffect, useRef, useState} from 'react';

export default function useErrorHandling<D = any>() {
  const [error, setError] = useState<any>(undefined);
  const mounted = useRef<boolean>();

  const mightFail = useCallback(async function callback<T = D, R = unknown>(
    retriever: Promise<T>,
    successHandler: (response: T) => R,
    errorHandler?: ((error: any) => void) | undefined,
    finallyHandler?: () => void
  ): Promise<R | undefined> {
    try {
      const response = await retriever;
      if (mounted.current) {
        return await successHandler(response);
      }
    } catch (error) {
      if (mounted.current) {
        errorHandler?.(error);
        setError(error);
      }
    } finally {
      finallyHandler?.();
    }
  }, []);

  const resetError = useCallback(() => {
    setError(undefined);
  }, []);

  useEffect(() => {
    mounted.current = true;
    return () => {
      mounted.current = false;
    };
  }, []);

  return {error, resetError, mightFail};
}
