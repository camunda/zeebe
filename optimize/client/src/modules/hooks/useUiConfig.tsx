/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useContext} from 'react';

import {UiConfig, configContext} from 'config';

export default function useUiConfig(): UiConfig {
  const contextValue = useContext(configContext);

  if (!contextValue) {
    throw new Error('useUiConfig has to be used within <ConfigProvider>');
  }

  return contextValue.config;
}
