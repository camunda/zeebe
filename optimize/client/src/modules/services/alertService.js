/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {del, get, post, put} from 'request';
import {track} from 'tracking';

export async function loadAlerts(collection) {
  const response = await get(`api/collection/${collection}/alerts`);
  return await response.json();
}

export async function addAlert(alert) {
  const response = await post(`api/alert`, alert);
  const json = await response.json();
  track('createAlert', {entityId: json.id});
  return response;
}

export async function editAlert(id, alert) {
  const response = await put(`api/alert/${id}`, alert);
  track('updateAlert', {entityId: id});
  return response;
}

export async function removeAlert(id) {
  const response = await del(`api/alert/${id}`);
  track('deleteAlert', {entityId: id});
  return response;
}
