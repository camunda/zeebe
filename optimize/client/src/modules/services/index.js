/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {
  getFlowNodeNames,
  checkDeleteConflict,
  loadDecisionDefinitionXml,
  loadInputVariables,
  loadOutputVariables,
  loadProcessDefinitionXml,
  loadVariables,
} from './dataLoaders';

export {numberParser} from './NumberParser';
export {incompatibleFilters} from './incompatibleFilters.ts';
export {default as reportConfig, createReportUpdate, getDefaultSorting} from './reportConfig';
export {getDiagramElementsBetween} from './diagramServices';
export {default as getTooltipText} from './getTooltipText';
export {default as getScreenBounds} from './getScreenBounds';
export {loadEntity, loadReports, updateEntity, deleteEntity, copyReport} from './entityService';
export {UNAUTHORIZED_TENANT_ID} from './tenantService.ts';
export {loadEntities, copyEntity, createEntity, getEntityIcon} from './entityService.tsx';

export {
  evaluateReport,
  isDurationReport,
  loadRawData,
  getReportResult,
  processResult,
  isAlertCompatibleReport,
  isCategoricalBar,
  isCategorical,
  TEXT_REPORT_MAX_CHARACTERS,
  isTextTileTooLong,
  isTextTileValid,
} from './reportService';

export {addSources, getCollection} from './collectionService';

export * as formatters from './formatters';

export function getRandomId() {
  return Math.random().toString(36).slice(2);
}

export function capitalize(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

export {loadAlerts, addAlert, removeAlert, editAlert} from './alertService';
export {default as ignoreFragments} from './ignoreFragments';
export {default as isReactElement} from './isReactElement';
export {loadDefinitions} from './loadDefinitions';
