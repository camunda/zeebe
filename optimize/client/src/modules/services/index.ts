/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export function getRandomId() {
  return Math.random().toString(36).slice(2);
}

export {loadReports, loadEntities, copyEntity, createEntity, getEntityIcon} from './entityService';
export {UNAUTHORIZED_TENANT_ID} from './tenantService';
export * as formatters from './formatters';
export {
  loadProcessDefinitionXml,
  loadDecisionDefinitionXml,
  loadInputVariables,
  loadOutputVariables,
  loadVariables,
} from './dataLoaders';
export {numberParser} from './NumberParser';
export {
  TEXT_REPORT_MAX_CHARACTERS,
  isTextTileTooLong,
  isTextTileValid,
  loadRawData,
  evaluateReport,
} from './reportService';
export {addSources, getCollection} from './collectionService';
export {default as getScreenBounds} from './getScreenBounds';
export {default as ignoreFragments} from './ignoreFragments';
export {default as isReactElement} from './isReactElement';
export {incompatibleFilters} from './incompatibleFilters';
export {loadDefinitions} from './loadDefinitions';

export type {Definition} from './loadDefinitions';
export type {ReportEvaluationPayload} from './reportService';
