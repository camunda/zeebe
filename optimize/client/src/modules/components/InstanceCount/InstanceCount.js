/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React, {useState} from 'react';
import equals from 'fast-deep-equal';

import {Popover} from 'components';
import {FilterList} from 'filter';
import {loadVariables, loadInputVariables, loadOutputVariables} from 'services';
import {t} from 'translation';
import {useErrorHandling} from 'hooks';
import {showError} from 'notifications';

import './InstanceCount.scss';

export default function InstanceCount({
  report,
  noInfo,
  additionalFilter,
  showHeader,
  trigger = (
    <Popover.Button size="sm" kind="ghost" className="defaultTrigger">
      {t('report.instanceCount.appliedFilters')}.
    </Popover.Button>
  ),
}) {
  const [hasLoaded, setHasLoaded] = useState(false);
  const [variables, setVariables] = useState();
  const {mightFail} = useErrorHandling();

  const {data, reportType} = report;

  const {key, versions, tenantIds} = data.definitions?.[0] ?? {};

  const instanceCount = report.result?.instanceCount;
  const totalCount = report.result?.instanceCountWithoutFilters;
  const hasFilter = data.filter?.length > 0;

  function loadRequiredData() {
    if (!key) {
      return;
    }

    if (reportType === 'process') {
      const payload = {
        processesToQuery: [
          {
            processDefinitionKey: key,
            processDefinitionVersions: versions,
            tenantIds: tenantIds,
          },
        ],
        filter: data.filter,
      };
      mightFail(loadVariables(payload), setVariables, showError);
    } else if (reportType === 'decision') {
      const payload = [
        {
          decisionDefinitionKey: key,
          decisionDefinitionVersions: versions,
          tenantIds: tenantIds,
        },
      ];
      mightFail(
        Promise.all([loadInputVariables(payload), loadOutputVariables(payload)]),
        ([inputVariable, outputVariable]) => setVariables({inputVariable, outputVariable}),
        showError
      );
    }
  }

  // for dashboards, we need to separate report level and dashboard level (additional) filters
  const reportFilters = [];
  const additionalFilters = [];

  if (hasFilter) {
    const unappliedAdditionalFilters = [...(additionalFilter ?? [])];

    data.filter.forEach((filter) => {
      const additionalFilterIdx = unappliedAdditionalFilters.findIndex((additionalFilter) =>
        equals(additionalFilter, filter)
      );
      if (additionalFilterIdx !== -1) {
        additionalFilters.push(filter);

        // Since the same filter can be applied twice (once as report level and once as dashboard level filter),
        // we keep track of which additional filters we already used. If a filter is twice in the data.filter array,
        // but only once in the additionalFilter array, we sort one copy to the reportFilters and one to the additionalFilters
        unappliedAdditionalFilters.splice(additionalFilterIdx, 1);
      } else {
        reportFilters.push(filter);
      }
    });
  }

  return (
    <div className="InstanceCount">
      {hasFilter && (
        <span
          onClick={() => {
            if (!noInfo && !hasLoaded) {
              // only load data if user actually opens the popover and
              // store the result to avoid loading data multiple times
              setHasLoaded(true);
              loadRequiredData();
            }
          }}
        >
          <Popover
            title={t('report.instanceCount.appliedFilters')}
            trigger={trigger}
            disabled={noInfo}
            className="instanceCountPopover"
            floating
          >
            {showHeader && (
              <div className="countString">
                {typeof instanceCount === 'number' &&
                  t(
                    `report.instanceCount.${reportType}.label${totalCount !== 1 ? '-plural' : ''}`,
                    {
                      count: instanceCount,
                      totalCount,
                    }
                  )}
              </div>
            )}
            {reportFilters.length > 0 && (
              <>
                {showHeader && (
                  <div className="filterListHeading">
                    {t('report.instanceCount.reportFiltersHeading')}
                  </div>
                )}
                <FilterList
                  definitions={data.definitions}
                  data={reportFilters}
                  variables={variables}
                  expanded
                />
              </>
            )}
            {additionalFilters.length > 0 && (
              <>
                <div className="filterListHeading">
                  {t('report.instanceCount.additionalFiltersHeading')}
                </div>
                <FilterList
                  definitions={data.definitions}
                  data={additionalFilters}
                  variables={variables}
                  expanded
                />
              </>
            )}
          </Popover>
        </span>
      )}
      <span className="countString">
        {typeof instanceCount === 'number' &&
          t(
            `report.instanceCount.${reportType}.label${totalCount !== 1 ? '-plural' : ''}${
              hasFilter ? '-withFilter' : ''
            }`,
            {
              count: instanceCount,
              totalCount: (haveDateFilter(data.filter) ? '*' : '') + totalCount,
            }
          )}
      </span>
    </div>
  );
}

function haveDateFilter(filters) {
  return filters?.some((filter) => filter.type.toLowerCase().includes('date'));
}
