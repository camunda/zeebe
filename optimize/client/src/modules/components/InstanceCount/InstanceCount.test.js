/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import update from 'immutability-helper';

import {loadVariables, loadInputVariables, loadOutputVariables} from 'services';
import {useErrorHandling} from 'hooks';

import InstanceCount from './InstanceCount';

jest.mock('services', () => {
  return {
    ...jest.requireActual('services'),
    loadInputVariables: jest
      .fn()
      .mockReturnValue([{id: 'input1', name: 'Input 1', type: 'String'}]),
    loadOutputVariables: jest
      .fn()
      .mockReturnValue([{id: 'output1', name: 'Output 1', type: 'String'}]),
    loadVariables: jest.fn().mockReturnValue([{name: 'variable1', type: 'String'}]),
  };
});

jest.mock('hooks', () => ({
  useErrorHandling: jest.fn(() => ({
    mightFail: jest.fn((data, cb) => cb(data)),
  })),
}));

beforeEach(() => {
  loadInputVariables.mockClear();
  loadOutputVariables.mockClear();
  loadVariables.mockClear();
});

const props = {
  report: {
    data: {
      filter: [{type: 'runningInstancesOnly'}],
      definitions: [
        {
          key: 'aKey',
          versions: ['1'],
          tenantIds: ['tenantId'],
        },
      ],
    },
    result: {
      instanceCount: 123,
      instanceCountWithoutFilters: 500,
    },
    reportType: 'process',
  },
};

it('should should show the instance count', () => {
  const node = shallow(<InstanceCount {...props} />);

  expect(node).toIncludeText('123');
});

it('should not show the total instance count if there are no filters', () => {
  const noFilterReport = update(props.report, {data: {filter: {$set: []}}});
  const node = shallow(<InstanceCount {...props} report={noFilterReport} />);

  expect(node).not.toIncludeText('500');
});

it('should show the total instance count if there are filters', () => {
  const node = shallow(<InstanceCount {...props} />);

  expect(node).toIncludeText('500');
});

it('should display an astrick near the total instance count if the report includes date filters', () => {
  const node = shallow(
    <InstanceCount
      {...props}
      report={update(props.report, {
        data: {filter: {$set: [{type: 'instanceStartDate'}]}},
      })}
    />
  );

  expect(node).toIncludeText('Displaying data from 123 of *500 instances');
});

it('should contain a popover with information about the filters', () => {
  const node = shallow(<InstanceCount {...props} />);

  expect(node.find('.instanceCountPopover')).toExist();
  expect(node.find('FilterList').prop('data')).toEqual(props.report.data.filter);
});

it('should separate report and dashboard level filters', () => {
  const allFilters = [
    {type: 'runningInstancesOnly', data: null},
    {
      type: 'processInstanceDuration',
      data: {value: 7, unit: 'days', operator: '>', includeNull: false},
    },
    {type: 'runningInstancesOnly', data: null},
    {type: 'nonCanceledInstancesOnly', data: null},
  ];

  const dashboardFilters = [
    {type: 'runningInstancesOnly', data: null},
    {type: 'nonCanceledInstancesOnly', data: null},
  ];

  const reportWithAllFilters = update(props.report, {data: {filter: {$set: allFilters}}});

  const node = shallow(
    <InstanceCount {...props} report={reportWithAllFilters} additionalFilter={dashboardFilters} />
  );

  expect(node.find('FilterList').at(0).prop('data')).toMatchSnapshot();
  expect(node.find('FilterList').at(1).prop('data')).toMatchSnapshot();
});

it('should not contain a popover if the report has no filters', () => {
  const noFilterReport = update(props.report, {data: {filter: {$set: []}}});
  const node = shallow(<InstanceCount {...props} report={noFilterReport} />);

  expect(node.find('.instanceCountPopover')).not.toExist();
});

it('should disable the popover if the noInfo prop is set', () => {
  const node = shallow(<InstanceCount {...props} noInfo />);

  expect(node.find('.instanceCountPopover').prop('disabled')).toBe(true);
});

it('should load variable names for process reports', async () => {
  const node = shallow(<InstanceCount {...props} />);

  node.find('span').first().simulate('click');

  await flushPromises();

  const payload = {
    processesToQuery: [
      {
        processDefinitionKey: 'aKey',
        processDefinitionVersions: ['1'],
        tenantIds: ['tenantId'],
      },
    ],
    filter: [{type: 'runningInstancesOnly'}],
  };

  expect(loadVariables).toHaveBeenCalledWith(payload);
  expect(node.find('FilterList').prop('variables')).toEqual([{name: 'variable1', type: 'String'}]);
});

it('should not load variables if definition is incomplete', async () => {
  const node = shallow(
    <InstanceCount {...props} report={update(props.report, {data: {definitions: {$set: []}}})} />
  );

  node.find('span').first().simulate('click');

  await flushPromises();

  expect(loadVariables).not.toHaveBeenCalled();
});

it('should load variable names for decision reports', async () => {
  useErrorHandling.mockReturnValueOnce({mightFail: async (data, cb) => cb(await data)});
  const decisionReport = update(props.report, {
    reportType: {$set: 'decision'},
  });
  const node = shallow(<InstanceCount {...props} report={decisionReport} />);

  node.find('span').first().simulate('click');

  await flushPromises();

  const payload = [
    {
      decisionDefinitionKey: 'aKey',
      decisionDefinitionVersions: ['1'],
      tenantIds: ['tenantId'],
    },
  ];

  expect(loadInputVariables).toHaveBeenCalledWith(payload);
  expect(loadOutputVariables).toHaveBeenCalledWith(payload);
  expect(node.find('FilterList').prop('variables')).toMatchSnapshot();
});

it('should use a custom trigger if passed', () => {
  const node = shallow(<InstanceCount {...props} trigger={<button />} />);

  expect(node.find('.instanceCountPopover').prop('trigger')).toEqual(<button />);
});

it('should show instance count and filter list headings if showHeader prop is added', () => {
  const node = shallow(<InstanceCount {...props} showHeader />);

  expect(node.find('.instanceCountPopover .countString')).toExist();
  expect(node.find('.filterListHeading')).toExist();
});
