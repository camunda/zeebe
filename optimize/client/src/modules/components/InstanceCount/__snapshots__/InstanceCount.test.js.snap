// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should load variable names for decision reports 1`] = `
Object {
  "inputVariable": Array [
    Object {
      "id": "input1",
      "name": "Input 1",
      "type": "String",
    },
  ],
  "outputVariable": Array [
    Object {
      "id": "output1",
      "name": "Output 1",
      "type": "String",
    },
  ],
}
`;

exports[`should separate report and dashboard level filters 1`] = `
Array [
  Object {
    "data": Object {
      "includeNull": false,
      "operator": ">",
      "unit": "days",
      "value": 7,
    },
    "type": "processInstanceDuration",
  },
  Object {
    "data": null,
    "type": "runningInstancesOnly",
  },
]
`;

exports[`should separate report and dashboard level filters 2`] = `
Array [
  Object {
    "data": null,
    "type": "runningInstancesOnly",
  },
  Object {
    "data": null,
    "type": "nonCanceledInstancesOnly",
  },
]
`;
